docs/RFCS/20210604_distributed_token_bucket.md

```diff
@@ -0,0 +1,416 @@
+and a sequence number which is unique within that tenant. The operation UUID
+allows detection of duplicate requests, to ensure idempotency. The sequence
+numbers allow locating the latest entry for each tenant.
+
+Schema for the system table:
+```sql
+CREATE TABLE tenant_usage (
+  tenant_id INT NOT NULL,
+  seq INT NOT NULL,
+
+  op_id UUID NOT NULL,  -- TODO: maybe INT?
+  op_time TIMESTAMP NOT NULL,
+
+  ru_burst_limit FLOAT NOT NULL,
+  ru_rate FLOAT NOT NULL,
+
+  ru_current FLOAT NOT NULL,
+  current_share_sum FLOAT NOT NULL,
+
+  -- Cumulative usage statistics.
+  total_ru_usage FLOAT NOT NULL,
+  total_read_requests INT NOT NULL,
+  total_read_bytes INT NOT NULL,
+  total_write_requests INT NOT NULL,
+  total_write_bytes INT NOT NULL,
+  total_sql_pod_cpu_seconds FLOAT NOT NULL, -- TODO: Maybe milliseconds and INT8?
+
+  PRIMARY KEY (tenant_id, seq DESC),
+  UNIQUE (tenant_id, op_id) -- could also be UNIQUE(op_id)
+);
+```
+
+An operation to the bucket is implemented with a transaction:
+```sql
+-- Generate random op_id.
+BEGIN;
+-- Get the latest state (with the largest sequence number).
+SELECT FOR UPDATE * FROM tenant_usage WHERE tenant_id=.. ORDER BY seq DESC LIMIT 1;
+-- Calculate new state.
+-- Set the new state, using the next sequence number.
+INSERT INTO tenant_usage (tenant_id, seq, ...) (tenant_id, seq+1, ...);
+END;
+```
+
+Any failed request is retried with the same `op_id`. If the previous request
+went through, the transaction will error out with a conflict on `op_id`,
+allowing special handling of this case.
+
+We perform periodic garbage collection to keep the table from growing. Each
+operation can trigger garbage collection (with random chance).
+
+```sql
+-- by number of entries
+DELETE FROM tenant_usage WHERE tenant_id=.. AND seq < (
+  SELECT seq - 1000 FROM tenant_usage WHERE tenant_id=.. ORDER BY seq DESC LIMIT 1
+)
+-- or by timestamp:
+DELETE FROM tenant_usage WHERE tenant_id=.. AND op_time < now() - '24hr'
+```
+
+TODO: it's unfortunate that we will have to talk to two ranges on each
+operation. Could we achieve this with a single index? We could reserve part of
+the UUID space for the sequence number, allowing one index to serve a
+dual-purpose, but things would become messy.
+
+Another possible variation would be to have a single `(tenant_id, op_id)`
+primary index and *also* store the current state and sequence number in
+`(tenant_id, '00000000-0000-0000-0000-000000000000')`. That would not require a
+secondary index, but would generate larger KVs as we store all fields twice.
+
+## KV Connector API
+
+The KV Connector API is used by the SQL pods to make requests to the distributed
+bucket. This API is implemented on the host cluster side by performing a
+transaction on `tenant_usage`.
+
+```go
+type TenantTokenBucketProvider interface {
+   Request(
+     ctx context.Context,
+     uuid util.UUID,
+     prevShareValue float64,
+     prevShareTimestamp time.Time,
+     tokensRequested float64,
```

Comment:
I know this isn't actual code, but a method with so many fields wouldn't get past CI, and even here it might be worth breaking it down using a few structs which will also condense this a bit.